src/ripple/app/consensus/RCLConsensus.cpp

```diff
@@ -633,29 +633,31 @@ RCLConsensus::Adaptor::doAccept(
         JLOG(j_.info()) << "We closed at "
                         << closeTime.time_since_epoch().count();
         using usec64_t = std::chrono::duration<std::uint64_t>;
-        usec64_t closeTotal =
-            std::chrono::duration_cast<usec64_t>(closeTime.time_since_epoch());
+        std::vector<usec64_t> closeTimes;
         int closeCount = 1;
 
         for (auto const& p : rawCloseTimes.peers)
         {
-            // FIXME: Use median, not average
             JLOG(j_.info())
                 << std::to_string(p.second) << " time votes for "
                 << std::to_string(p.first.time_since_epoch().count());
+
             closeCount += p.second;
-            closeTotal += std::chrono::duration_cast<usec64_t>(
-                              p.first.time_since_epoch()) *
-                p.second;
+
+            auto time = std::chrono::duration_cast<usec64_t>(p.first.time_since_epoch());
+            for(int pi = 0; pi < p.second; ++pi)
+              closeTimes.push_back(time);
         }
 
-        closeTotal += usec64_t(closeCount / 2);  // for round to nearest
-        closeTotal /= closeCount;
+        auto our_time = std::chrono::duration_cast<usec64_t>(closeTime.time_since_epoch());
+        closeTimes.insert(std::upper_bound(closeTimes.begin(), closeTimes.end(), our_time), our_time);
 
         // Use signed times since we are subtracting
         using duration = std::chrono::duration<std::int32_t>;
         using time_point = std::chrono::time_point<NetClock, duration>;
-        auto offset = time_point{closeTotal} -
+
+        // Use median time
+        auto offset = time_point{closeTimes[closeTimes.size()/2]} -
```

Comment:
This change will, unfortunately, make the code more flakey in a very subtle and hard to detect way: it relies on an implementation detail (namely that `rawCloseTimes.peers` is an ordered container) without documenting this.

I remain unconvinced that the median (especially the weighted median) is the right metric here. Why not use a simpler and more easily understood metric of central tendency? For example by picking the most popular close time and resolving conflicts by the earliest time. While in theory such a metric could be less robust than the median, I think in practice it would be fine, especially if we improve the dampening/hysteresis in `adjustCloseTime`.

Pinging Messrs. @HowardHinnant and @JoelKatz.

Reply:
Hey @nbougalis, thanks for the feedback as well, I hit reply before I saw your comment. I'm curious as to more details to the "most popular" close time idea, would this entail selecting the time with >= 51% of the votes, or simply the time w/ the highest percentage of votes (and would there be a minimum threshold enforced, eg what if the most popular option only got 25% of the votes).

Reply:
@movitto Thanks for looking at this. I agree with Nik that median may not be the best for this purpose and is probably better to do holistically by considering how we adjust the network clock based on this distribution of peer close times. I suspect that in practice, median and mean and almost anything else aren't terribly different, but it would be useful to see what the distribution does look like.

If that is out of scope for this PR, I'd be vote removing the FIXME in the code and opening an issue instead to track this.

Reply:
@nbougalis @bachase Hey Nik, Brad sorry this past week has been hectic, haven't been able to get to it yet, but I am planning on following up on this ASAP!